core/auth/src/main/java/software/amazon/awssdk/auth/signer/internal/AbstractAws4Signer.java

```diff
@@ -333,52 +317,21 @@ private void addPreSignInformationToRequest(SdkHttpFullRequest.Builder mutableRe
 
         mutableRequest.putRawQueryParameter(SignerConstant.X_AMZ_ALGORITHM, SignerConstant.AWS4_SIGNING_ALGORITHM);
         mutableRequest.putRawQueryParameter(SignerConstant.X_AMZ_DATE, signerParams.getFormattedRequestSigningDateTime());
-        mutableRequest.putRawQueryParameter(SignerConstant.X_AMZ_SIGNED_HEADER, signedHeadersString);
+        mutableRequest.putRawQueryParameter(SignerConstant.X_AMZ_SIGNED_HEADER, canonicalRequest.signedHeaderString());
         mutableRequest.putRawQueryParameter(SignerConstant.X_AMZ_EXPIRES, Long.toString(expirationInSeconds));
         mutableRequest.putRawQueryParameter(SignerConstant.X_AMZ_CREDENTIAL, signingCredentials);
     }
 
-    private Map<String, List<String>> canonicalizeSigningHeaders(Map<String, List<String>> headers) {
-        Map<String, List<String>> result = new TreeMap<>();
-
-        for (Map.Entry<String, List<String>> header : headers.entrySet()) {
-            String lowerCaseHeader = lowerCase(header.getKey());
-            if (LIST_OF_HEADERS_TO_IGNORE_IN_LOWER_CASE.contains(lowerCaseHeader)) {
-                continue;
-            }
-
-            result.computeIfAbsent(lowerCaseHeader, x -> new ArrayList<>()).addAll(header.getValue());
-        }
-
-        return result;
-    }
-
-    private String getCanonicalizedHeaderString(Map<String, List<String>> canonicalizedHeaders) {
-        StringBuilder buffer = new StringBuilder();
-
-        canonicalizedHeaders.forEach((headerName, headerValues) -> {
-            buffer.append(headerName);
-            buffer.append(":");
-            buffer.append(String.join(",", trimAll(headerValues)));
-            buffer.append("\n");
-        });
-
-        return buffer.toString();
-    }
-
-    private List<String> trimAll(List<String> values) {
-        return values.stream().map(HeaderTransformsHelper::trimAll).collect(Collectors.toList());
-    }
-
-    private String getSignedHeadersString(Map<String, List<String>> canonicalizedHeaders) {
-        StringBuilder buffer = new StringBuilder();
-        for (String header : canonicalizedHeaders.keySet()) {
-            if (buffer.length() > 0) {
-                buffer.append(";");
-            }
-            buffer.append(header);
-        }
-        return buffer.toString();
+    /**
+     * Tests a char to see if is it whitespace.
+     * This method considers the same characters to be white
+     * space as the Pattern class does when matching \s
+     *
+     * @param ch the character to be tested
+     * @return true if the character is white  space, false otherwise.
+     */
+    private boolean isWhiteSpace(final char ch) {
+        return ch == ' ' || ch == '\t' || ch == '\n' || ch == '\u000b' || ch == '\r' || ch == '\f';
```

Comment:
Guava's charmatcher allows denser checks, with a binary search and binary sets, which might be handy with 6 + chars to check for.... Especially if in the future you decide to add more.

Reply:
I considered doing something like a 256 byte array lookup table for chars being whitespace (with some special handling for unicode), but this method didn't show up very high on a profiler so it felt like premature optimization. It would be nice if we could have a Guava dependency, but since we have to implement it ourselves (or deal with importing it), I wasn't inclined.